File: lib/rgfa/line/header/multiline.rb
#
# Support for headers defined in multiple lines of the GFA file. The lines may
# also contain the same tag defined multiple times in different lines, with
# multiple values.
#
module RGFA::Line::Header::Multiline

  # Tags which are allowed to be defined with a single value in the header
  # (if multiple header line contain them, they must contain the same value)
  SINGLE_DEFINITION_TAGS = [:VN, :TS]

  # Set a header tag value.
  #
  # If the tag +tagname+ does not exist yet, it is set to +value+.  If it exists
  # and is defined a single time (i.e. it is not a field array): if the tagname
  # is in the {SINGLE_DEFINITION_TAGS} list, if value is the same as before, it
  # is ignored, if different, an error is raised; for other tags: a field array
  # is created, containing the previous value and +value+. If the previous value
  # it is a field array (i.e. the tag was already defined multiple times),
  # +value+ is added to it.
  #
  # @param tagname [Symbol] a valid GFA custom or predefined tag name
  # @param value [Object] the value to which the tag shall be set; if the tag
  #   was already defined (or if a datatype parameter is specified),
  #   the value must be compatible with the datatype
  # @param datatype [RGFA::Field::TAG_DATATYPE, nil] the GFA tag datatype to
  #   use; if none is specified, it is determined by the previous values for
  #   the field (if any) or by the class of value (otherwise)
  #
  # @raise [RGFA::FormatError] if the tagname is invalid
  # @raise [RGFA::TypeError, RGFA::FormatError]
  #    if the value is not compatible with the specified datatype
  # @raise [RGFA::InconsistencyError]
  #    if the value is not compatible with the datatype of the previously
  #    defined value; or a tag in the {SINGLE_DEFINITION_TAGS} list is defined
  #    multiple times with different values.
  #
  # @return [self]
  def add(tagname, value, datatype=nil)
    tagname = tagname.to_sym
    prev = get(tagname)
    if prev.nil?
      set_datatype(tagname, datatype) if datatype
      set(tagname, value)
      return self
    elsif !prev.kind_of?(RGFA::FieldArray)
      if SINGLE_DEFINITION_TAGS.include?(tagname)
        if field_to_s(tagname) == value.to_gfa_field(fieldname: tagname)
          return self
        else
          raise RGFA::InconsistencyError,
            "Inconsistent values for header tag #{tagname} found\n"+
            "Previous definition: #{prev}\n"+
            "Current definition: #{value}"
        end
      end
      prev = RGFA::FieldArray.new(get_datatype(tagname), [prev])
      set_existing_field(tagname, prev)
    end
    if @vlevel > 1
      prev.vpush(value, datatype, tagname)
    else
      prev << value
    end
    return self
  end

  # Compute the string representation of a header field.
  #
  # @param fieldname [Symbol] the tag name of the field
  # @param tag [Boolean] <i>(defaults to: +false+)</i>
  #   return the tagname:datatype:value representation
  #
  # @raise [RGFA::NotFoundError] if field is not defined
  #
  # @return [String] the string representation
  def field_to_s(fieldname, tag: false)
    prev = get(fieldname)
    if prev.kind_of?(RGFA::FieldArray)
      prev.validate_gfa_field(nil, fieldname) if @vlevel >= 2
      return tag ? prev.to_gfa_tag(fieldname)
                 : prev.to_gfa_field(fieldname: fieldname)
    else
      super
    end
  end

  # @api private
  module API_PRIVATE

    # Count the tags which are represented by FieldArray (i.e. with multiple
    # definitions).
    #
    # @return [Integer (>= 0)]
    def n_duptags
      n = 0
      tagnames.each do |tn|
        n += 1 if get(tn).kind_of?(RGFA::FieldArray)
      end
      return n
    end

    # Split the header line into single-tag lines.
    #
    # If a tag is a FieldArray, this is splitted into multiple fields with the
    # same fieldname (note that it leads to an invalid GFA, if all instances
    # are output in the same GFA line).
    #
    # @return [Array<RGFA::Line::Header>]
    def split
      tags.map do |tagname, datatype, value|
        h = RGFA::Line::Header.new([], vlevel: @vlevel)
        h.set_datatype(tagname, datatype)
        h.set(tagname, value)
        h
      end
    end

    # Merge an additional {RGFA::Line::Header} line into this header line.
    #
    # @param gfa_line [RGFA::Line::Header] the header line to merge
    #
    # @return [self]
    def merge(gfa_line)
      gfa_line.tagnames.each do |of|
        add(of, gfa_line.get(of), gfa_line.get_datatype(of))
      end
      self
    end

  end
  include API_PRIVATE

  private

  # Array of tags data.
  #
  # Returns the tags as an array of [fieldname, datatype, value] arrays. If a
  # field is a FieldArray, this is splitted into multiple fields with the same
  # fieldname.
  #
  # @return [Array<(Symbol, Symbol, Object)>]
  def tags
    retval = []
    tagnames.each do |of|
      value = get(of)
      if value.kind_of?(RGFA::FieldArray)
        value.each do |elem|
          retval << [of, value.datatype, elem]
        end
      else
        retval << [of, get_datatype(of), value]
      end
    end
    return retval
  end

end

File: gfapy/line/header/multiline.py
import gfapy

class Multiline:
  """
  Implementation of the support for multiple header line in the
  GFA file (which also may contain the same value defined multiple
  times in different lines).
  """

  SINGLE_DEFINITION_TAGS = ["VN", "TS"]

  def add(self, tagname, value, datatype = None):
    """
    Set a header value (multi-value compatible).

    If a field does not exist yet, set it to value. If it exists and it is a
    *gfapy.FieldArray*, add the value to the field array. If it exists and it
    is not a field array, create a field array with the previous value and
    the new one.

    Parameters
    ----------
    tagname : str
    value : object
    datatype : gfapy.Field.TAG_DATATYPE, optional
      The datatype to use.
      The default is to determine the datatype according to the value or the
      previous values present in the field.
    """
    prev = self.get(tagname)
    if prev is None:
      if datatype is not None:
        self.set_datatype(tagname, datatype)
      self.set(tagname, value)
      return
    elif not isinstance(prev, gfapy.FieldArray):
      if tagname in self.SINGLE_DEFINITION_TAGS:
        if self.field_to_s(tagname) == \
            gfapy.Field._to_gfa_field(value, fieldname=tagname):
          return
        else:
          raise gfapy.InconsistencyError(
            "Inconsistent values for header tag {} found\n".format(tagname)+
            "Previous definition: {}\n".format(prev)+
            "Current definition: {}".format(value))
      prev = gfapy.FieldArray(self.get_datatype(tagname), [prev])
      self._set_existing_field(tagname, prev)
    if self.vlevel > 1:
      prev.vpush(value, datatype, tagname)
    else:
      prev.append(value)

  def field_to_s(self, fieldname, tag = False):
    """
    Compute the string representation of a field.

    Parameters
    ----------
    fieldname : str
      The tag name of the field.
    tag : bool
      *(defaults to: ***False***)*
      Return the 'tagname:datatype:value' representation.

    Raises
    ------
    gfapy.NotFoundError
      If field is not defined.

    Returns
    -------
    str
      The string representation.
    """
    prev = self.get(fieldname)
    if isinstance(prev, gfapy.FieldArray):
      if self.vlevel >= 2:
        prev._validate_gfa_field(None, fieldname)
      return prev._to_gfa_tag(fieldname=fieldname) if tag else \
             prev._to_gfa_field(fieldname=fieldname)
    else:
      return super(gfapy.line.header.Line, self).field_to_s(fieldname, tag)

  def _n_duptags(self):
    n = 0
    for tn in self.tagnames:
      if isinstance(self.get(tn),gfapy.FieldArray):
        n+=1
    return n

  def _split(self):
    """
    Split the header line into single-tag lines.

    If a tag is a FieldArray, this is splitted into multiple fields
    with the same fieldname.

    Returns
    -------
    gfapy.line.Header list
    """
    retval = []
    for tagname, datatype, value in self._tags():
      h = gfapy.line.Header(["H"], vlevel = self.vlevel)
      h.set_datatype(tagname, datatype)
      h.set(tagname, value)
      retval.append(h)
    return retval

  def _merge(self, gfa_line):
    """
    Merge an additional **gfa.line.Header** line into this header line.

    Parameters
    ----------
    gfa_line : gfapy.line.Header
      The header line to merge.

    Returns
    -------
    self
    """
    for of in gfa_line.tagnames:
      self.add(of, gfa_line.get(of), gfa_line.get_datatype(of))
    return self

  def _tags(self):
    """
    List of tags data.

    Returns the tags as an list of [fieldname, datatype, value]
    lists. If a field is a FieldArray, this is splitted into multiple fields
    with the same fieldname.

    Returns
    -------
    (str, str, object) list
    """
    retval = []
    for of in self.tagnames:
      value = self.get(of)
      if isinstance(value, gfapy.FieldArray):
        for elem in value:
          retval.append((of, value.datatype, elem))
      else:
        retval.append((of, self.get_datatype(of), value))
    return retval
